executor/common_linux.h

```diff
@@ -1464,6 +1464,23 @@ static long syz_extract_tcp_res(volatile long a0, volatile long a1, volatile lon
 #include "common_usb_linux.h"
 #endif
 
+#if SYZ_EXECUTOR || __NR_syz_raw_bind_info
+#include <linux/raw.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
+
+static long syz_raw_bind_info(volatile long inptr, volatile long raw_minor, volatile long outptr)
```

Comment:
This pseudo-syscall will not work, but is also necessary now.
It will not work because the fuzzer engine only passes resources across syscall, it does not pass whole structs or somehow arrange that a struct at the same address is reused across syscalls. So what the current descriptions say the fuzzer to do is something along the following lines:
```
raw_bind_info foo;
syz_raw_bind_info(stat, 1, &foo);

raw_bind_info bar = {0x42, 0x43, 0x44}; // some random numbers as descriptions say to fill
ioctl$RAW_CHAR_CTRL_SETBIND(fd, RAW_SETBIND, &bar);
```
There is no relation between foo and bar in any way.
There is also no relation between stat syscall output and the stat struct that we accept into syz_raw_bind_info, so the fuzzer will just put random numbers into the stat struct that is passed to syz_raw_bind_info.

But now that the syz_raw_bind_info is very simple and merely copies values, that's specifically what resources are for. So we if declared dev major/minor in stat as output resources and specify them as input resources for raw_bind_info struct, then the fuzzer will do what syz_raw_bind_info attempts to do auto-magically.

But we can do this even simpler and more efficient.
You initial PR only bound to loop devices (and generally stat can be applied to lots of different types of files and majority of them are not block devices) and major for the loop device is a known const:
https://elixir.bootlin.com/linux/v5.8-rc4/source/include/uapi/linux/major.h#L22
and minors are also known, they are simply 0, 1, ....
So we can do this:
```
raw_bind_info {
	raw_minor	proc[1, 2, int32]
	block_major	const[LOOP_MAJOR, int64] # potentially we can use flags and add other suitable major types
	block_minor	proc[0, 1, int64]
}
```

Reply:
I see, would it be a good idea to cross reference the list in that link with [Documentation/admin-guide/devices.txt](https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt), or just keep it with loop major only?
This way we could cover at least all devices that have a known const as major.
I'd only have to specify multiple structs to keep major+minor numbers correct, but that would work.

A question I have is how that works with multiple processes in a vm. I know loop devices are nicely separated ([from discussion on groups](https://groups.google.com/d/msg/syzkaller/NE2vRYGaYSA/oWfAyfHkBgAJ)), but would this work nicely with other devices at all?

Reply:
Adding more device types is good, I think it can give some new interesting coverage.
We don't want all of them b/c some are not present/esoteric/require hardware, and some are dangerous. In my qemu I see:

```
root@syzkaller:~# ls -l /dev/block/
total 0
lrwxrwxrwx 1 root root  6 Jul 16 06:59 11:0 -> ../sr0
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:0 -> ../ram0
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:1 -> ../ram1
lrwxrwxrwx 1 root root  8 Jul 16 06:59 1:10 -> ../ram10
lrwxrwxrwx 1 root root  8 Jul 16 06:59 1:11 -> ../ram11
lrwxrwxrwx 1 root root  8 Jul 16 06:59 1:12 -> ../ram12
lrwxrwxrwx 1 root root  8 Jul 16 06:59 1:13 -> ../ram13
lrwxrwxrwx 1 root root  8 Jul 16 06:59 1:14 -> ../ram14
lrwxrwxrwx 1 root root  8 Jul 16 06:59 1:15 -> ../ram15
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:2 -> ../ram2
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:3 -> ../ram3
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:4 -> ../ram4
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:5 -> ../ram5
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:6 -> ../ram6
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:7 -> ../ram7
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:8 -> ../ram8
lrwxrwxrwx 1 root root  7 Jul 16 06:59 1:9 -> ../ram9
lrwxrwxrwx 1 root root  9 Jul 16 06:59 250:0 -> ../nullb0
lrwxrwxrwx 1 root root  8 Jul 16 06:59 252:0 -> ../zram0
lrwxrwxrwx 1 root root  8 Jul 16 06:59 259:0 -> ../pmem0
lrwxrwxrwx 1 root root 12 Jul 16 06:59 31:0 -> ../mtdblock0
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:0 -> ../nbd0
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:128 -> ../nbd4
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:160 -> ../nbd5
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:192 -> ../nbd6
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:224 -> ../nbd7
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:256 -> ../nbd8
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:288 -> ../nbd9
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:32 -> ../nbd1
lrwxrwxrwx 1 root root  8 Jul 16 06:59 43:320 -> ../nbd10
lrwxrwxrwx 1 root root  8 Jul 16 06:59 43:352 -> ../nbd11
lrwxrwxrwx 1 root root  8 Jul 16 06:59 43:384 -> ../nbd12
lrwxrwxrwx 1 root root  8 Jul 16 06:59 43:416 -> ../nbd13
lrwxrwxrwx 1 root root  8 Jul 16 06:59 43:448 -> ../nbd14
lrwxrwxrwx 1 root root  8 Jul 16 06:59 43:480 -> ../nbd15
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:64 -> ../nbd2
lrwxrwxrwx 1 root root  7 Jul 16 06:59 43:96 -> ../nbd3
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:0 -> ../loop0
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:1 -> ../loop1
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:2 -> ../loop2
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:3 -> ../loop3
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:4 -> ../loop4
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:5 -> ../loop5
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:6 -> ../loop6
lrwxrwxrwx 1 root root  8 Jul 16 06:59 7:7 -> ../loop7
lrwxrwxrwx 1 root root  6 Jul 16 06:59 8:0 -> ../sda
lrwxrwxrwx 1 root root  6 Jul 16 06:59 9:0 -> ../md0
```

sr0 is I think cd-rom, we can use it.
ram0/zram0 are dangerous to use, it will smash machine memory.
nullb0 is good to use.
pmem0  is good to use.
nbd0  is good to use.
mtdblock0 and md0 I don't know.
sda is dangerous to use, it will smash root disk.

Reply:
> A question I have is how that works with multiple processes in a vm. I know loop devices are nicely separated (from discussion on groups), but would this work nicely with other devices at all?

It won't nicely with other devices. If a device is global and is not suited for proper isolated testing, then there is not much we can do.
loop devices are also not isolated perfectly b/c we have this, which I think can open all of them:
syz_open_dev$loop(dev ptr[in, string["/dev/loop#"]], id intptr, flags flags[open_flags]) fd_loop
We could use proc type here as wel. Or, better switch to use of loopfs which allows proper isolation.